### PR TITLE
Add connected peers to protocol, use in sync provider

### LINKED ISSUE #1853: Remove blocking operations in SyncProvider
The current implementation of `SyncProvider` has two methods that block awaiting a reply from another thread. 

https://github.com/paritytech/substrate/blob/1ca4cc0a16a357782bb1028bb57376594ca232a0/core/network/src/service.rs#L279

and 

https://github.com/paritytech/substrate/blob/1ca4cc0a16a357782bb1028bb57376594ca232a0/core/network/src/service.rs#L287

`status` seems to be mostly used by the `informant` at https://github.com/paritytech/substrate/blob/d4eb4bac4fc66d54f429a953170139525bd5c7ee/core/cli/src/informant.rs#L51

while `peers` is used by the rpc system at https://github.com/paritytech/substrate/blob/b5587b378af80027a056a10e77bbf1ddd8b04d15/core/rpc/src/system/mod.rs#L113

It would appear that `status` could be refactored as a stream of status updates similar to the https://github.com/paritytech/substrate/blob/d4eb4bac4fc66d54f429a953170139525bd5c7ee/core/cli/src/informant.rs#L115 
used by the informant, so that it's not the informant asking for a status and waiting for the reply, instead it would be network providing a stream of status updates to be handled by the informant. 

`peers` looks more complicated to refactor, especially since it involves creating a reply from two different threads(`Protocol` and the libp2p `ServiceTask`). The question would be "how to we provide updates with regards to connected peers without having the other end having to ask for it and wait for the reply?".

I can imagine a stream of updates, with the end-user being able to fast-forward to the latest update, or handle them one at a time and potentially not use the latest update. 

One could also use a lock, if that would be more fine-grained then using a reply channel, and just looking at `Protocol`, it uses `context_data` and the `peers` it contains almost every time it does something, and also passes it as `ProtocolContext` to specialization, gossip, and so on. So it would appear that the other end could end-up waiting almost as much on the lock as it would on the reply from the channel.  

We could maybe do something with https://github.com/crossbeam-rs/crossbeam/blob/master/crossbeam-skiplist/src/set.rs

### PR DESCRIPTION
**Pull request by gterzian**
Fix #1853

### CODE REVIEW

**Comment by gterzian** [comment]
The idea behind this change is that we can use `context_data.peers` as a thread-local piece of data, that is used extensively by the protocol thread. Then we have `connected_peers` that is shared with other threads via `SyncProvider`, and that is updated only when necessary by the protocol thread. This should reduce blocking on awaiting the lock on `connected_peers` to a minimum.

File: core/network/src/test/mod.rs [hunk 1]
```diff
@@ -181,7 +184,8 @@ impl<D> Peer<D> {
 
 	/// Called on connection to other indicated peer.
 	fn on_connect(&self, other: NodeIndex) {
-		let _ = self.protocol_sender.send(ProtocolMsg::PeerConnected(other, String::new()));
+		let peer_id = PublicKey::Ed25519((0 .. 32).collect::<Vec<u8>>()).into();
```
Comment (tomaka) [inline core/network/src/test/mod.rs]: There's a `PeerId::random()` method which may be more explicit.

File: core/rpc/src/system/tests.rs [hunk 2]
```diff
@@ -41,8 +41,9 @@ impl network::SyncProvider<Block> for Status {
 		}
 	}
 
-	fn peers(&self) -> Vec<(NodeIndex, Option<PeerId>, NetworkPeerInfo<Block>)> {
-		vec![(1, Some(PublicKey::Ed25519((0 .. 32).collect::<Vec<u8>>()).into()), NetworkPeerInfo {
+	fn peers(&self) -> Vec<(NodeIndex, NetworkPeerInfo<Block>)> {
+		vec![(1, NetworkPeerInfo {
+			peer_id: Some(PublicKey::Ed25519((0 .. 32).collect::<Vec<u8>>()).into()),
```
Comment (tomaka) [inline core/rpc/src/system/tests.rs]: Same remark: `PeerId::random()`.

File: core/network/src/protocol.rs [hunk 3]
```diff
@@ -743,25 +789,20 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
...
-					.send(NetworkMsg::GetPeerId(who.clone(), sender));
-				let node_id = port.recv().expect("1. We are running 2. Network should be running too.");
-				if let Some(id) = node_id {
+				if let Some(id) = peer.info.peer_id.as_ref() {
 					for hash in hashes {
 						propagated_to
 							.entry(hash)
 							.or_insert_with(Vec::new)
-							.push(id.clone());
+							.push(id.to_base58());
```
Comment (tomaka) [inline core/network/src/protocol.rs]: Why this change? `PeerId` implements `Hash`, so it should be possible to insert it.

Comment (gterzian) [reply core/network/src/protocol.rs]: The code seems to be expecting a `String`. Note that this isn't really a change because previously the call to `to_base58` was done at https://github.com/paritytech/substrate/pull/1857/files#diff-a045635f3eec17c06b3c7ac7944acceaL522

File: core/network/src/service.rs [hunk 4]
```diff
@@ -284,16 +288,9 @@ impl<B: BlockT + 'static, S: NetworkSpecialization<B>> SyncProvider<B> for Servi
...
-		let peers = port.recv().expect("1. Protocol keeps handling messages until all senders are dropped,
-			or the ProtocolMsg::Stop message is received,
-			2 Service keeps a sender to protocol, and the ProtocolMsg::Stop is never sent.");
-		let network = self.network.lock();
-		peers.into_iter().map(|(idx, info)| {
-			(idx, network.peer_id_of_node(idx).map(|p| p.clone()), info)
-		}).collect::<Vec<_>>()
+	fn peers(&self) -> Vec<(NodeIndex, PeerInfo<B>)> {
+		let peers = (*self.peers.read()).clone();
+		peers.into_iter().map(|(idx, connected)| (idx, connected.peer_info)).collect::<Vec<_>>()
```
Comment (twittner) [inline core/network/src/service.rs]:
```suggestion
		peers.into_iter().map(|(idx, connected)| (idx, connected.peer_info)).collect()
```

File: core/network/src/protocol.rs [hunk 5]
```diff
@@ -685,21 +703,49 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
...
+						.write()
+						.insert(who, ConnectedPeer { peer_id, peer_info: peer_info.clone() });
+					peer_info
+				},
+				None => {
+					// Note: can this actually happen?
+					debug!(target: "sync", "Received status from previously unconnected node {}", who);
+					let peer_id = None;
+					let peer_info = PeerInfo {
+						peer_id: peer_id.clone(),
```
Comment (twittner) [inline core/network/src/protocol.rs]:
```suggestion
						peer_id: None,
```

**Comment by gterzian** [comment]
Comments addressed, and rebased. Sorry, it's out of habit!

**Comment by gterzian** [comment]
Ok, had to reshuffle the rpc tests a bit to be able to use `PeerId::random()`...

**Comment by gterzian** [comment]
@tomaka @twittner Your comments have been addressed, and I had to make a few additional changes to the rpc test code.

File: core/network/src/protocol.rs [hunk 6]
```diff
@@ -66,9 +67,27 @@ pub struct Protocol<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> {
...
-	handshaking_peers: HashMap<NodeIndex, time::Instant>,
+	handshaking_peers: HashMap<NodeIndex, HandshakingPeer>,
+	// Connected peers from whom we received a Status message,
+	// similar to context_data.peers but shared with the SyncProvider.
+	connected_peers: Arc<RwLock<HashMap<NodeIndex, ConnectedPeer<B>>>>,
 	transaction_pool: Arc<TransactionPool<H, B>>,
 }
+
+#[derive(Clone)]
+/// A peer from whom we have received a Status message.
```
Comment (andresilva) [inline core/network/src/protocol.rs]: style: docs before attributes

File: core/network/src/protocol.rs [hunk 7]
```diff
@@ -303,12 +324,8 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
 
 	fn handle_msg(&mut self, msg: ProtocolMsg<B, S>) -> bool {
 		match msg {
-			ProtocolMsg::Peers(sender) => {
-				let peers = self.context_data.peers.iter().map(|(idx, p)| (*idx, p.info.clone())).collect();
-				let _ = sender.send(peers);
-			},
-			ProtocolMsg::PeerDisconnected(who, debug_info) => self.on_peer_disconnected(who, debug_info),
-			ProtocolMsg::PeerConnected(who, debug_info) => self.on_peer_connected(who, debug_info),
+			ProtocolMsg::PeerDisconnected(who, debug_info) => self.on_peer_disconnected( who, debug_info),
```
Comment (andresilva) [inline core/network/src/protocol.rs]:
```suggestion
			ProtocolMsg::PeerDisconnected(who, debug_info) => self.on_peer_disconnected(who, debug_info),
```

File: core/network/src/protocol.rs [hunk 8]
```diff
@@ -676,21 +694,48 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
...
 			let cache_limit = NonZeroUsize::new(1_000_000).expect("1_000_000 > 0; qed");
 
-			let peer = Peer {
-				info: PeerInfo {
-					protocol_version: status.version,
-					roles: status.roles,
-					best_hash: status.best_hash,
-					best_number: status.best_number
+			let info = match self.handshaking_peers.remove(&who) {
+				Some(hanshaking) => {
```
Comment (andresilva) [inline core/network/src/protocol.rs]:
```suggestion
				Some(handshaking) => {
```

File: core/network/src/protocol.rs [hunk 9]
```diff
@@ -676,21 +694,48 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
...
+					let peer_id = Some(hanshaking.peer_id);
```
Comment (andresilva) [inline core/network/src/protocol.rs]:
```suggestion
					let peer_id = Some(handshaking.peer_id);
```

File: core/network/src/protocol.rs [hunk 10]
```diff
@@ -66,9 +67,27 @@ pub struct Protocol<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> {
...
+	// Connected peers from whom we received a Status message,
+	// similar to context_data.peers but shared with the SyncProvider.
+	connected_peers: Arc<RwLock<HashMap<NodeIndex, ConnectedPeer<B>>>>,
 	transaction_pool: Arc<TransactionPool<H, B>>,
 }
+
+#[derive(Clone)]
+/// A peer from whom we have received a Status message.
+pub struct ConnectedPeer<B: BlockT> {
+	pub peer_id: Option<PeerId>,
```
Comment (andresilva) [inline core/network/src/protocol.rs]: This should not be an option, since we go from handshake -> connected and we already know the peer id on the handshake. I guess we can even remove this field if it's also not defined as `Option` in `PeerInfo`.

File: core/network/src/protocol.rs [hunk 11]
```diff
@@ -99,6 +118,8 @@ struct Peer<B: BlockT, H: ExHashT> {
 /// Info about a peer's known state.
 #[derive(Clone, Debug)]
 pub struct PeerInfo<B: BlockT> {
+	/// Network id.
+	pub peer_id: Option<PeerId>,
```
Comment (andresilva) [inline core/network/src/protocol.rs]: As above, should not be an option.

File: core/network/src/protocol.rs [hunk 12]
```diff
@@ -676,21 +694,48 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
...
+						roles: status.roles,
+						best_hash: status.best_hash,
+						best_number: status.best_number
+					};
+					self.connected_peers
+						.write()
+						.insert(who, ConnectedPeer { peer_id, peer_info: peer_info.clone() });
+					peer_info
+				},
+				None => {
```
Comment (andresilva) [inline core/network/src/protocol.rs]: This statement should be unreachable so I think we should just log the error and return.

**Comment by andresilva** [comment]
Nice one @gterzian :+1:
